Add pull request workflow

HarrisonWAffel opened the pull request:
## Problem
#1351 missed a workflow for pull requests, as such we are not building and packaging the provider before allowing PRs to be merged 
 
## Solution
Add a pull request workflow which builds and packages the provider for each PR. This PR does not include the acceptance tests, as additional effort needs to be put into how we can modernize those tests such that they can work against a more modern version of rancher. 
 
## Testing
I've tested this workflow on my personal fork 

Actions run: https://github.com/HarrisonWAffel/terraform-provider-rancher2/actions/runs/9571691853/job/26389359180

snasovich reviewed (commented):
LGTM at a glance but some observations/questions:
1. On Drone we ran `make docker-build` and `make build-rancher` is basically equivalent of it minus Docker. Why do we need to run `make package-rancher`?
2. Any specific reason we're running on `ubuntu-20.04` vs. `ubuntu-latest`?

HarrisonWAffel commented:
I've updated to `ubuntu-latest`, and removed `make package-rancher`. Initial thought process was to more closely mirror what release CI does, but if this step was not being done while running in drone I agree we should drop it.

jiaqiluo reviewed (approved):
LGTM

snasovich commented on .github/workflows/pull_request.yaml:
```diff
@@ -0,0 +1,16 @@
+  pull_request:
+
+jobs:
+  buildAndPackage:
```
Hate to nit but we should rename it to just `build` if we removed packaging step.

HarrisonWAffel replied:
Good catch

HarrisonWAffel committed:
Add pull request workflow

snasovich reviewed (approved):
LGTM as soon as https://github.com/rancher/terraform-provider-rancher2/actions/runs/9572629031/job/26392362041?pr=1359 succeeds
